channels/rdpdr/client/rdpdr_main.c

```diff
@@ -649,63 +637,127 @@ static BOOL isAutomountLocation(const char* path)
 	return FALSE;
 }
 
-/**
- * Function description
- *
- * @return 0 on success, otherwise a Win32 error code
- */
-static UINT handle_hotplug(rdpdrPlugin* rdpdr)
+#define MAX_USB_DEVICES 100
+
+typedef struct _hotplug_dev
 {
-	FILE* f;
-	hotplug_dev dev_array[MAX_USB_DEVICES] = { 0 };
-	size_t i;
-	size_t size = 0;
-	int count, j;
+	char* path;
+	BOOL to_add;
+} hotplug_dev;
+
+static void
+handle_mountpoint(hotplug_dev *dev_array, size_t *size, const char *mountpoint)
+{
+	if (!mountpoint)
+		return;
+	/* copy hotpluged device mount point to the dev_array */
+	if (isAutomountLocation(mountpoint) && (*size < MAX_USB_DEVICES))
+	{
+		dev_array[*size].path = _strdup(mountpoint);
+		dev_array[*size + 1].to_add = TRUE;
+		(*size)++;
+	}
+}
+
 #ifdef __sun
+#include <sys/mnttab.h>
+static UINT
+handle_platform_mounts_sun(hotplug_dev *dev_array, size_t *size)
+{
+	FILE* f;
 	struct mnttab ent;
-#else
-	struct mntent* ent;
+	f = fopen("/etc/mnttab", "r");
+	if (f == NULL)
+	{
+		WLog_ERR(TAG, "fopen failed!");
+		return ERROR_OPEN_FAILED;
+	}
+	while (getmntent(f, &ent) == 0)
+	{
+		handle_mountpoint(dev_array, size, ent.mnt_mountp);
+	}
+	fclose(f);
+	return ERROR_SUCCESS;
+}
 #endif
-	ULONG_PTR* keys = NULL;
-	UINT32 ids[1];
-	UINT error = 0;
 
-#ifdef __sun
-	f = fopen("/etc/mnttab", "r");
-#else
-	f = fopen("/proc/mounts", "r");
+#if defined(__FreeBSD__) || defined(__OpenBSD__)
+#include <sys/mount.h>
+static UINT
+handle_platform_mounts_bsd(hotplug_dev *dev_array, size_t *size)
+{
+	int mntsize;
+	size_t idx;
+	struct statfs *mntbuf = NULL;
+
+	mntsize = getmntinfo(&mntbuf, MNT_NOWAIT);
+	if (!mntsize)
+	{
+		/* TODO: handle 'errno' */
+		WLog_ERR(TAG, "getmntinfo failed!");
+		return ERROR_OPEN_FAILED;
+	}
+	for (idx = 0; idx < (size_t)mntsize; idx ++)
+	{
+		handle_mountpoint(dev_array, size, mntbuf[idx].f_mntonname);
+	}
+	free(mntbuf);
+	return ERROR_SUCCESS;
+}
 #endif
 
+#if defined(__LINUX__) || defined (__linux__)
+#include <mntent.h>
+static UINT
+handle_platform_mounts_linux(hotplug_dev *dev_array, size_t *size)
+{
+	FILE* f;
+	struct mntent* ent;
+	f = fopen("/proc/mounts", "r");
 	if (f == NULL)
 	{
 		WLog_ERR(TAG, "fopen failed!");
 		return ERROR_OPEN_FAILED;
 	}
-
-#ifdef __sun
-	while (getmntent(f, &ent) == 0)
-#else
 	while ((ent = getmntent(f)) != NULL)
-#endif
 	{
-		/* Copy the line, path must obviously be shorter */
+		handle_mountpoint(dev_array, size, ent->mnt_dir);
+	}
+	fclose(f);
+	return ERROR_SUCCESS;
+}
+#endif
+
+static UINT
+handle_platform_mounts(hotplug_dev *dev_array, size_t *size)
```

Comment:
This can be removed, just name all the platform specific functions `handle_platform_mounts`